monai/data/dataset.py

```diff
@@ -93,22 +97,64 @@ def __init__(self, data, transform, cache_num=sys.maxsize, cache_rate=1.0):
                 will take the minimum of (cache_num, data_length x cache_rate, data_length).
             cache_rate (float): percentage of cached data in total, default is 1.0 (cache all).
                 will take the minimum of (cache_num, data_length x cache_rate, data_length).
+            num_workers (int): the number of worker processes to use.
+                If 0 a single thread will be used. Default is 0.
         """
         if not isinstance(transform, Compose):
             transform = Compose(transform)
         super().__init__(data, transform)
         self.cache_num = min(cache_num, int(len(self) * cache_rate), len(self))
-        self._cache = list()
+        self._cache = [None] * self.cache_num
         print('Load and cache transformed data...')
-        for i in range(self.cache_num):
-            process_bar(i + 1, self.cache_num)
-            item = data[i]
-            for _transform in transform.transforms:
-                # execute all the deterministic transforms before the first random transform
-                if isinstance(_transform, Randomizable):
-                    break
-                item = apply_transform(_transform, item)
-            self._cache.append(item)
+        if num_workers > 0:
+            self._item_processed = 0
+            if mode == "thread":
+                lock = Lock()
+
+                def _update_bar(arg):
+                    self._item_processed += 1
+                    lock.acquire()
+                    try:
+                        process_bar(self._item_processed, self.cache_num)
+                    finally:
+                        lock.release()
+                with ThreadPoolExecutor(max_workers=num_workers) as executor:
+                    for i in range(self.cache_num):
+                        future = executor.submit(self._load_cache_item_thread, i, transform, data)
+                        future.add_done_callback(_update_bar)
+            elif mode == "process":
+                def _add_item(res):
+                    item, i = res
+                    self._cache[i] = item
+                    self._item_processed += 1
+                    process_bar(self._item_processed, self.cache_num)
+                with Pool(num_workers) as p:
+                    for i in range(self.cache_num):
+                        item = data[i]
+                        p.apply_async(self._load_cache_item_process, args=(i, transform, item), callback=_add_item)
```

Comment:
This is going to apply `self._load_cache_item_process` in one process at a time, perhaps you should use `Pool.map` instead to broadcast out the work to multiple processes at once and let the pool manage job assignment?

Reply:
Good idea, I'll try it, thanks.